clang/test/CodeGenSYCL/group-local-memory.cpp

```diff
@@ -1,29 +0,0 @@
-// Check that SYCLLowerWGLocalMemory pass is added to the SYCL device
-// compilation pipeline with the inliner pass (new Pass Manager).
-
-// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown -emit-llvm -O2 \
-// RUN:   -mdebug-pass Structure %s -o /dev/null 2>&1 \
-// RUN:   | FileCheck %s -check-prefixes=CHECK-INL,CHECK
-
-// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown -emit-llvm -O0 \
-// RUN:   -mdebug-pass Structure %s -o /dev/null 2>&1 \
-// RUN:   | FileCheck %s --check-prefixes=CHECK-ALWINL,CHECK
-
-// Check that AlwaysInliner pass is always run for compilation of SYCL device
-// target code, even if all optimizations are disabled.
```

Comment:
According to [the comment](https://github.com/intel/llvm/blob/35fb5064e8686050cde25ff4d5e5cdf7fb8ac43c/llvm/include/llvm/SYCLLowerIR/LowerWGLocalMemory.h#L8-L9), it's important to run the inliner pass before `SYCLLowerWGLocalMemoryPass`. It enables the check that `__sycl_allocateLocalMemory` is called in the kernel scope only. Unfortunately, this check [hasn't been implemented](https://github.com/intel/llvm/blob/35fb5064e8686050cde25ff4d5e5cdf7fb8ac43c/llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp#L55-L57).

> In addition, intel gpu compiler has a pass to transform global variable in addrspace(3) to alloca that runs after pipeline basic simplification.

Wouldn't it better to put `SYCLLowerWGLocalMemoryPass` at the end of "basic simplification pipeline"?

Reply:
@bader does SYCL forbids declaration of static local memory in non-kernel function?
My test shows cuda support this.

Reply:
https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_local_memory.asciidoc#allocating-local-memory
As of now it's a requirement.

Reply:
https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_local_memory.asciidoc#allocating-local-memory also mentions `NOTE: The restriction that group-local variables must be defined at kernel functor scope may be lifted in a future version of this extension.`

> Wouldn't it better to put `SYCLLowerWGLocalMemoryPass` at the end of "basic simplification pipeline"?

Inliner pass run a relatively late stage after "basic simplification pipeline", so above change won't solve the issue that inliner is required.

> it's important to run the inliner pass before SYCLLowerWGLocalMemoryPass. It enables the check that __sycl_allocateLocalMemory is called in the kernel scope only.

ModuleInlinerWrapperPass runs on O2 but not on O0. Therefore, putting SYCLLowerWGLocalMemoryPass at the end of the pipeline doesn't resolve the inlining issue.
Is it ok to add alwaysinline to callers of __sycl_allocateLocalMemory recursively so that it will be in kernel scope?

Reply:
To be honest, I don't an easy way to detect if `__sycl_allocateLocalMemory` was called in non-kernel scope. Let's go with your patch "as is". We have already TODO in the code.